cpp/include/raft/comms/std_comms.hpp

```diff
@@ -0,0 +1,406 @@
+    initialize();
+    p2p_enabled = true;
+  };
+
+  /**
+   * @brief constructor for collective-only operation
+   * @param comm initilized nccl communicator
+   * @param size size of the cluster
+   * @param rank rank of the current worker
+   */
+  std_comms(ncclComm_t comm, int size, int rank)
+    : _nccl_comm(comm), _size(size), _rank(rank) {
+    initialize();
+  };
+
+  virtual ~std_comms() {
+    CUDA_CHECK_NO_THROW(cudaStreamDestroy(_stream));
+
+    CUDA_CHECK_NO_THROW(cudaFree(_sendbuff));
+    CUDA_CHECK_NO_THROW(cudaFree(_recvbuff));
+  }
+
+  void initialize() {
+    CUDA_CHECK(cudaStreamCreate(&_stream));
+
+    CUDA_CHECK(cudaMalloc(&_sendbuff, sizeof(int)));
+    CUDA_CHECK(cudaMalloc(&_recvbuff, sizeof(int)));
+  }
+
+  int getSize() const { return _size; }
+
+  int getRank() const { return _rank; }
+
+  std::unique_ptr<comms_iface> commSplit(int color, int key) const {
+    // Not supported by NCCL
+    ASSERT(false,
+           "ERROR: commSplit called but not yet supported in this comms "
+           "implementation.");
+  }
+
+  void barrier() const {
+    CUDA_CHECK(cudaMemsetAsync(_sendbuff, 1, sizeof(int), _stream));
+    CUDA_CHECK(cudaMemsetAsync(_recvbuff, 1, sizeof(int), _stream));
+
+    allreduce(_sendbuff, _recvbuff, 1, INT, SUM, _stream);
+
+    ASSERT(syncStream(_stream) == status_t::commStatusSuccess,
+           "ERROR: syncStream failed. This can be caused by a failed rank.");
+  }
+
+  void get_request_id(request_t *req) const {
+    request_t req_id;
+
+    if (this->_free_requests.empty())
+      req_id = this->_next_request_id++;
+    else {
+      auto it = this->_free_requests.begin();
+      req_id = *it;
+      this->_free_requests.erase(it);
+    }
+    *req = req_id;
+  }
+
+  void isend(const void *buf, int size, int dest, int tag,
+             request_t *request) const {
+    ASSERT(_ucp_worker != nullptr,
+           "ERROR: UCX comms not initialized on communicator.");
+
+    get_request_id(request);
+    ucp_ep_h ep_ptr = (*_ucp_eps)[dest];
+
+    ucp_request *ucp_req = (ucp_request *)malloc(sizeof(ucp_request));
+
+    this->_ucp_handler.ucp_isend(ucp_req, ep_ptr, buf, size, tag,
+                                 default_tag_mask, getRank());
+
+    _requests_in_flight.insert(std::make_pair(*request, ucp_req));
+  }
+
+  void irecv(void *buf, int size, int source, int tag,
+             request_t *request) const {
+    ASSERT(_ucp_worker != nullptr,
+           "ERROR: UCX comms not initialized on communicator.");
+
+    get_request_id(request);
+
+    ucp_ep_h ep_ptr = (*_ucp_eps)[source];
+
+    ucp_tag_t tag_mask = default_tag_mask;
+
+    ucp_request *ucp_req = (ucp_request *)malloc(sizeof(ucp_request));
+    _ucp_handler.ucp_irecv(ucp_req, _ucp_worker, ep_ptr, buf, size, tag,
+                           tag_mask, source);
+
+    _requests_in_flight.insert(std::make_pair(*request, ucp_req));
+  }
+
+  void waitall(int count, request_t array_of_requests[]) const {
```

Comment:
This synchronizes only on specific point-to-point messages (per the MPI interface) while `barrier()` synchronizes the entire cluster.

Reply:
waitall is wait on array_of_requests[] (you get a request for each non-blocking communication function call), this is different from communication barrier.

Reply:
Like the rest of the comms API, waitall is taken directly from the MPI API and is meant to synchronize for multiple point to point messages (which are always non-blocking in our API). We did this because we can implement blocking versions easily with non-blocking methods but the reverse is not true.